source/extensions/access_loggers/http_grpc/grpc_access_log_impl.cc

```diff
@@ -32,64 +32,85 @@ TLSProperties_TLSVersion tlsVersionStringToEnum(const std::string& tls_version)
 
-void GrpcAccessLogStreamerImpl::ThreadLocalStreamer::send(
-    envoy::service::accesslog::v2::StreamAccessLogsMessage& message, const std::string& log_name) {
-  auto stream_it = stream_map_.find(log_name);
-  if (stream_it == stream_map_.end()) {
-    stream_it = stream_map_.emplace(log_name, ThreadLocalStream(*this, log_name)).first;
+void GrpcAccessLoggerImpl::log(envoy::data::accesslog::v2::HTTPAccessLogEntry& entry) {
+  // TODO(euroelessar): Add batching and flushing.
+  envoy::service::accesslog::v2::StreamAccessLogsMessage message;
+  message.mutable_http_logs()->add_log_entry()->Swap(&entry);
+
+  if (stream_ == absl::nullopt) {
+    stream_.emplace(*this);
   }
 
-  auto& stream_entry = stream_it->second;
-  if (stream_entry.stream_ == nullptr) {
-    stream_entry.stream_ =
+  if (stream_->stream_ == nullptr) {
+    stream_->stream_ =
         client_->start(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
                            "envoy.service.accesslog.v2.AccessLogService.StreamAccessLogs"),
-                       stream_entry);
+                       *stream_);
 
     auto* identifier = message.mutable_identifier();
-    *identifier->mutable_node() = shared_state_->local_info_.node();
-    identifier->set_log_name(log_name);
+    *identifier->mutable_node() = local_info_.node();
+    identifier->set_log_name(log_name_);
   }
 
-  if (stream_entry.stream_ != nullptr) {
-    stream_entry.stream_->sendMessage(message, false);
+  if (stream_->stream_ != nullptr) {
+    stream_->stream_->sendMessage(message, false);
   } else {
     // Clear out the stream data due to stream creation failure.
-    stream_map_.erase(stream_it);
+    stream_.reset();
   }
 }
 
+GrpcAccessLoggerCacheImpl::GrpcAccessLoggerCacheImpl(Grpc::AsyncClientManager& async_client_manager,
+                                                     Stats::Scope& scope,
+                                                     ThreadLocal::SlotAllocator& tls,
+                                                     const LocalInfo::LocalInfo& local_info)
+    : async_client_manager_(async_client_manager), scope_(scope), tls_slot_(tls.allocateSlot()),
+      local_info_(local_info) {
+  tls_slot_->set([](Event::Dispatcher&) {
+    return ThreadLocal::ThreadLocalObjectSharedPtr{new ThreadLocalCache()};
+  });
+}
+
+GrpcAccessLoggerSharedPtr GrpcAccessLoggerCacheImpl::getOrCreateLogger(
+    const envoy::config::accesslog::v2::CommonGrpcAccessLogConfig& config) {
+  auto& cache = tls_slot_->getTyped<ThreadLocalCache>();
+  const std::size_t cache_key = MessageUtil::hash(config);
+  const auto it = cache.access_loggers_.find(cache_key);
+  if (it != cache.access_loggers_.end()) {
+    return it->second;
+  }
+  const Grpc::AsyncClientFactoryPtr factory =
+      async_client_manager_.factoryForGrpcService(config.grpc_service(), scope_, false);
+  const auto logger = std::static_pointer_cast<GrpcAccessLogger>(
```

Comment:
nit: you shouldn't need a pointer case here.

Reply:
Used explicit type instead with implicit conversion.